File: haskell/collatz-conjecture/src/CollatzConjecture.hs
module CollatzConjecture (collatz) where

collatz :: Integer -> Maybe Integer
collatz n
  | n < 1 = Nothing
  | otherwise = Just (accum n 0)
  where
    accum x c
      | x == 1 = c
      | even x = accum (x `div` 2) (c + 1)
      | otherwise = accum (x * 3 + 1) (c + 1)

File: c/collatz-conjecture/collatz_conjecture.c
#include "collatz_conjecture.h"

int steps_count(int n, int s);

int steps_count(int n, int s) {
    if (n == 1) return s;

    if (n%2 == 0) // even
        return steps_count(n/2, ++s);
    
    return steps_count(3*n + 1, ++s);
}


int steps(int start) {
    if (start <= 0) return ERROR_VALUE;

    return steps_count(start, 0);
}

